### PR TITLE
Adding DataFrame for the RPacket Tracking Functionality

### PR DESCRIPTION
**Pull request by DhruvSondhi**
This is in continuation of the #1748 
Adds the functionality of creating a dataframe from the interactions values of the packets tracked.
This PR will be rebased once #1748 is merged leading to addition of only a few commits.

**Description**

**Motivation and context**

**How has this been tested?**
- [x] Testing pipeline.
- [ ] Other. 

**Examples**

**Type of change**

- [ ] Bug fix. 
- [x] New feature. 
- [ ] Breaking change. 
- [ ] None of the above. 

**Checklist**

- [x] My change requires a change to the documentation.
    - [x] I have updated the documentation accordingly.
    - [ ] (optional) I have built the documentation on my fork following [the instructions](https://tardis-sn.github.io/tardis/development/documentation_preview.html).
- [x] I have assigned and requested two reviewers for this pull request.

Link to Documentation : https://dhruvsondhi.github.io/tardis/branch/packet_interaction_dataframe/io/output/rpacket_tracking.html

### CODE REVIEW

**Comment by DhruvSondhi** [comment]
This PR needs to be rebased after #1748 has been merged.

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 1]
```diff
@@ -50,6 +52,13 @@ def montecarlo_radial1d(model, plasma, runner):
         runner.j_blue_estimator,
         runner.Edotlu_estimator,
     )
+
+    # Configuring the Tracking for R_Packets
```
Comment (DhruvSondhi) [inline tardis/montecarlo/montecarlo_numba/base.py:56]: This is causing the Multithreading to fail. Investigation needed

Comment (Rodot-) [reply tardis/montecarlo/montecarlo_numba/base.py:65]: Has this been resolved?

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 2]
```diff
@@ -124,6 +134,18 @@ def montecarlo_radial1d(
...
+            iteration,
+        )
+
+        runner.rpacket_tracker = track_rpacket_dataframe(
+            runner.rpacket_tracker, tracked_df
+        )
+
 
 @njit(**njit_dict)
 def montecarlo_main_loop(
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:151]: Please update the docstring to include the returned tracks

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 3]
```diff
@@ -328,4 +364,75 @@ def montecarlo_main_loop(
...
         virt_packet_last_line_interaction_in_id,
         virt_packet_last_line_interaction_out_id,
+        rpacket_collections,
+    )
+
+
+def create_tracked_rpacket_df(rpacket_collections, iteration):
+    for i in range(len(rpacket_collections)):
+        if i == 0:
+            seed = rpacket_collections[0].seed
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:374]: Put this above the loop and iterate from 1 to len(rpacket_collections)

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 4]
```diff
@@ -328,4 +364,75 @@ def montecarlo_main_loop(
...
+            index = rpacket_collections[0].index
+            status = rpacket_collections[0].status
+            r = rpacket_collections[0].r
+            nu = rpacket_collections[0].nu
+            mu = rpacket_collections[0].mu
+            energy = rpacket_collections[0].energy
+            shell_id = rpacket_collections[0].shell_id
+
+        seed = [*seed, *rpacket_collections[i].seed]
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:383]: Look into `itertools.chain`

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 5]
```diff
@@ -328,4 +364,75 @@ def montecarlo_main_loop(
...
+
+    rpacket_tracked_df = pd.DataFrame(
+        zip(*vals), columns=columns_name, dtype=object
+    )
+
+    rpacket_tracked_df["Iteration"] = iteration
+    return rpacket_tracked_df
+
+
+def track_rpacket_dataframe(tracked_rpacket_df, tracked_df):
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:421]: Add docstring

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 1 above)
Comment (DhruvSondhi) [reply tardis/montecarlo/montecarlo_numba/base.py:65]: Yes, this has been resolved.

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 4 above)
Comment (DhruvSondhi) [reply tardis/montecarlo/montecarlo_numba/base.py:383]: Thank you for this. This works as intended :)

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 3 above)
Comment (DhruvSondhi) [reply tardis/montecarlo/montecarlo_numba/base.py:374]: I have made it such that it will explicitly create a array with the values from the `0` index and then append the other outputs to the previously created array. Thus, increasing the array dynamically.

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 5 above)
Comment (DhruvSondhi) [reply tardis/montecarlo/montecarlo_numba/base.py:434]: Added as required

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 6]
```diff
@@ -328,4 +364,75 @@ def montecarlo_main_loop(
         virt_packet_last_interaction_type,
         virt_packet_last_line_interaction_in_id,
         virt_packet_last_line_interaction_out_id,
+        rpacket_collections,
+    )
+
+
+def create_tracked_rpacket_df(rpacket_collections, iteration):
```
[the remark this answers is not in the document]
Comment (DhruvSondhi) [reply tardis/montecarlo/montecarlo_numba/base.py:369]: Added as required

**Comment by DhruvSondhi** [comment]
Why is this PR closed?

**Comment by wkerzendorf** [comment]
I don't think this tests running numba in parallel

File: docs/io/output/index.rst [hunk 7]
```diff
@@ -12,5 +12,6 @@ In addition to the widgets, TARDIS can output information in several other forms
     to_hdf
     callback
     vpacket_logging
-    progress_bars
     rpacket_tracking
+    progress_bars
```
Comment (wkerzendorf) [inline docs/io/output/index.rst:16]: okay this shouldn't be part of the PR, right?

File: docs/io/output/tardis_tracking_example.yml [hunk 8]
```diff
@@ -0,0 +1,57 @@
+# Example YAML configuration for TARDIS
```
Comment (wkerzendorf) [inline docs/io/output/tardis_tracking_example.yml:1]: this feels like the standard example?

File: docs/io/output/index.rst (hunk 7 above)
Comment (DhruvSondhi) [reply docs/io/output/index.rst:16]: This was rebased. I will check this.

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 9]
```diff
@@ -347,5 +366,86 @@ def montecarlo_main_loop(
...
+        "Packet Index",
+        "Packet Seed",
+        "Packet Status",
+        "r",
+        "nu",
+        "mu",
+        "energy",
+        "shell_id",
+    ]
+
```
Comment (Rodot-) [inline tardis/montecarlo/montecarlo_numba/base.py:400]:
```suggestion
size=0
for i in range(len(rpacket_collections)):
    size += rpacket_collections[i].interation_id

... allocate arrays of size `size`        
pos = 0
for i in range(len(rpacket_collections)):
    ... arrays[pos:pos+rpacket_collections[i].interaction_id] = rpacket_collections[i].whatever[::]
    pos += rpacket_collections[i].interactions_id
    
    
```

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 10]
```diff
@@ -347,5 +366,86 @@ def montecarlo_main_loop(
...
+    for i in range(len(rpacket_collections)):
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/base.py:401]:
```suggestion
    for rpt in rpacket_collections:
```

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 11]
```diff
@@ -347,5 +366,86 @@ def montecarlo_main_loop(
...
+            val_list = [
+                rpacket_collections[i].index[j],
+                rpacket_collections[i].seed[j],
+                rpacket_collections[i].status[j],
+                rpacket_collections[i].r[j],
+                rpacket_collections[i].nu[j],
+                rpacket_collections[i].mu[j],
+                rpacket_collections[i].energy[j],
+                rpacket_collections[i].shell_id[j],
+            ]
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/base.py:412]:
```suggestion
        for attribute in ['status', 'r', 'nu', 'mu', 'energy', 'shell_id']:
            rpacket_tracker_all[attribute].append(getattr(rpacket_tracker, attribute))
```

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 12]
```diff
@@ -347,5 +366,86 @@ def montecarlo_main_loop(
...
+
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/base.py:413]:
```suggestion
         rpacket_tracker_all['index'].append(np.ones_like(rpacket_tracker.shell_id)*rpacket_tracker.index)
         rpacket_tracker_all['seed'].append(np.ones_like(rpacket_tracker.shell_id)*rpacket_tracker.seed)
```

**Comment by wkerzendorf** [comment]
change to `RPacketTracking`

**Comment by DhruvSondhi** [comment]
> change to `RPacketTracker`

Addressed via #1861

File: docs/io/output/tardis_tracking_example.yml (hunk 8 above)
Comment (DhruvSondhi) [reply docs/io/output/tardis_tracking_example.yml:1]: This is not a standard example. There is changes to the number of packets so as to make it run fast for the tracking. More changes would be done in the future so as to make it very fast & not consume much time in testing.

Comment (wkerzendorf) [reply docs/io/output/tardis_tracking_example.yml:1]: let's change this after

File: tardis/io/tests/data/tardis_configv1_verysimple_tracking.yml [hunk 13]
```diff
@@ -0,0 +1,60 @@
+# Example YAML configuration for TARDIS
```
Comment (wkerzendorf) [inline tardis/io/tests/data/tardis_configv1_verysimple_tracking.yml:1]: same comment as above

File: tardis/io/logger/tests/test_logging.py [hunk 14]
```diff
@@ -105,3 +108,112 @@ def test_logging_both_specified(
...
+        track_df.to_hdf(tracker_ref_path, key=key, mode="w")
+        pytest.skip("Reference data was generated during this run.")
+
+
+@pytest.fixture
+def read_comparison_refdata(tracker_ref_path):
+    return pd.read_hdf(tracker_ref_path)
+
+
+# @pytest.mark.parametrize(
```
Comment (andrewfullard) [inline tardis/io/logger/tests/test_logging.py:149]: Clean out these comments

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
-from numba import prange, njit, jit, objmode
 import logging
 import numpy as np
+import pandas as pd
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/base.py:3]: I don't think this import, and perhaps some others, are being used any more

File: tardis/montecarlo/tests/test_montecarlo.py [hunk 16]
```diff
@@ -53,8 +53,10 @@
 import tardis.montecarlo.montecarlo_numba.utils as utils
```
Comment (andrewfullard) [inline tardis/montecarlo/tests/test_montecarlo.py:53]: Changes to this file should be reverted, they are formatting only.

**Review by andrewfullard — CHANGES_REQUESTED** [review]

**Comment by andrewfullard** [comment]
Closing, see #2073
